app/models/observation.rb

```diff
@@ -193,6 +238,186 @@ class Observation < AbstractModel
+        where.not(name: name_ids)
+    else
+      where(name_id: name_ids)
+    end
+  }
+  scope :of_name_like,
+        ->(name) { where(name: Name.text_name_includes(name)) }
+  scope :by_user,
+        ->(user) { where(user: user) }
+  scope :with_location,
+        -> { where.not(location: nil) }
+  scope :without_location,
+        -> { where(location: nil) }
+  scope :at_location,
+        ->(location) { where(location: location) }
+  scope :in_region,
+        ->(where) { where(Observation[:where].matches("%#{where}")) }
+  scope :in_box, # Use named parameters (n, s, e, w), any order
+        lambda { |**args|
+          if args[:s]&.between?(0, 90) && args[:n]&.between?(0, 90) &&
+             args[:w]&.between?(-180, 180) && args[:e]&.between?(-180, 180) &&
+             args[:s] <= args[:n] &&
+             ((args[:w] <= args[:e]) || (args[:w] >= 0 && args[:e] <= 0))
+
+            # expand box by epsilon to create leeway for Float rounding
+            # Fixes a bug where Califoria fixture was not in a box
+            # defined by the fixture's north, south, east, west
+            epsilon = 0.00001
+            n = args[:n] + epsilon
+            s = args[:s] - epsilon
+            e = args[:e] + epsilon
+            w = args[:w] - epsilon
+
+            # Does box straddle 180 deg?
+            if args[:e] < args[:w]
+              where(
+                # Observation[:long] between w & 180 OR between 180 and e
+                (Observation[:lat] >= s).
+                and(Observation[:lat] <= n).
+                and(Observation[:long] >= w).
+                or(Observation[:long] <= e)
+              )
+            else
+              where(
+                (Observation[:lat] >= s).
+                and(Observation[:lat] <= n).
+                and(Observation[:long] >= w).
+                and(Observation[:long] <= e)
+              )
+            end
+          else
+            Observation.none
+          end
+        }
+  scope :is_collection_location,
+        -> { where(is_collection_location: true) }
+  scope :not_collection_location,
+        -> { where(is_collection_location: false) }
+  scope :with_image,
+        -> { where.not(thumb_image: nil) }
+  scope :without_image,
+        -> { where(thumb_image: nil) }
+  scope :with_notes,
+        -> { where.not(notes: Observation.no_notes) }
+  scope :without_notes,
+        -> { where(notes: Observation.no_notes) }
+  scope :has_notes_field,
+        ->(field) { where(Observation[:notes].matches("%:#{field}:%")) }
+  scope :notes_include,
+        ->(notes) { where(Observation[:notes].matches("%#{notes}%")) }
+  scope :with_specimen,
+        -> { where(specimen: true) }
+  scope :without_specimen,
+        -> { where(specimen: false) }
+  scope :with_sequence,
+        -> { joins(:sequences).distinct }
+  scope :without_sequence, lambda {
+    # much faster than `missing(:sequences)` which uses left outer join.
```

Comment:
@nimmolo & @pellaea
Fyi we should probably avoid `missing` where possible. It uses a left outer join, which is slow.
Here's how `without_sequence` performs using `missing` vs the actual implementation.

| query |  user  |  system   |  total  |  real  |
| ----- | ---- | ------ | ----- | ---- | 
| missing   |   0.046535 | 0.014053 | 0.060588 | (0.069965) |
| where.not | 0.012196 | 0.000586 | 0.012782 | (0.013454) |

Reply:
Clever.  Does AR implement this as a subquery?  Or does it literally slurp all the sequences up and calculate the set of all observations with sequences?  I wouldn't be surprised if even that would be faster than the left outer join!

Reply:
Great question. It's the latter. I hadn't thought of doing it as a subquery; I don't know anything about them.
It's interesting how bad left outer joins are; even my crappy code is better.
You question also reveals that I'm too clever by half. It's muuuch much simpler and transparent as:
```ruby
where.not(id: with_sequence)
```
Now it's easy to see how it works:
```ruby
Observation.without_sequence.to_sql
 => "SELECT `observations`.* FROM `observations` WHERE `observations`.`id` NOT IN (SELECT DISTINCT `observations`.`id` FROM `observations` INNER JOIN `sequences` ON `sequences`.`observation_id` = `observations`.`id`)" 
```
I'll push that up in a minute. 
I don't have time to investigate subqueries tonight.

Reply:
That is exactly the subquery I was talking about.  Brilliant!  ActiveRecord really has its shit together.  That's very nicely done, I have to say.

Reply:
Oh! So that's a "subquery". I imagined it was a special sql function. That again proves that I don't know what I'm talking about.
It's too bad that `missing` doesn't figure out on its own that it can use a subquery instead of a left outer join.

Reply:
Yes, that is a little unfortunate.  But there may be cases where the left outer join is preferable(?)  This gives you the control to decide which you want.